src/groupeddataframe/grouping.jl

```diff
@@ -181,6 +197,319 @@ end
+6-element Array{DataFrames.GroupKey,1}:
+ GroupKey(a = :foo, b = 2)
+ GroupKey(a = :bar, b = 1)
+ GroupKey(a = :baz, b = 2)
+ GroupKey(a = :foo, b = 1)
+ GroupKey(a = :bar, b = 2)
+ GroupKey(a = :baz, b = 1)
+
+julia> k = first(keys(gd));
+
+julia> keys(k)
+(:a, :b)
+
+julia> values(k)
+(:foo, 2)
+
+julia> NamedTuple(k)
+(a = :foo, b = 2)
+
+julia> k.a
+:foo
+
+julia> k[:a]
+:foo
+
+julia> k[1]
+:foo
+
+julia> gd[k]
+2×3 SubDataFrame
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+```
+"""
+function Base.keys(gd::GroupedDataFrame)
+    return (GroupKey(gd, i) for i in 1:length(gd))
+end
+
+# Get group by GroupKey
+function Base.getindex(gd::GroupedDataFrame, key::GroupKey)
+    gd === parent(key) && return gd[getfield(key, :idx)]
+    throw(ErrorException("Cannot use a GroupKey to index a GroupedDataFrame other than the one it was derived from."))
+end
+
+function Base.getindex(gd::GroupedDataFrame, key::Tuple)
+    for (i, v) in enumerate(groupvalues(Tuple, gd))
+        isequal(v, key) && return gd[i]
+    end
+    throw(KeyError(key))
+end
+
+function Base.getindex(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
+    if length(key) != length(gd.cols) || any(n != _names(gd)[c] for (n, c) in zip(N, gd.cols))
+        throw(KeyError(key))
+    end
+    return gd[Tuple(key)]
+end
+
+"""
+    get(gd::GroupedDataFrame, key, default)
+
+Get a group based on the values of the grouping columns.
+
+`key` may be a `NamedTuple` or `Tuple` of grouping column values, or one of the
+[`GroupKey`](@ref)s returned by [`keys(gd::GroupedDataFrame)`](@ref).
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[2]),
+                      b = repeat([2, 1], outer=[3]),
+                      c = 1:6);
+
+julia> gd = groupby(df, :a)
+GroupedDataFrame with 3 groups based on key: a
+First Group (2 rows): a = :foo
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 1     │ 4     │
+⋮
+Last Group (2 rows): a = :baz
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 2     │ 3     │
+│ 2   │ baz    │ 1     │ 6     │
+
+julia> get(gd, (a=:bar,), nothing)
+2×3 SubDataFrame
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ bar    │ 1     │ 2     │
+│ 2   │ bar    │ 2     │ 5     │
+
+julia> get(gd, (:baz,), nothing)
+2×3 SubDataFrame
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 2     │ 3     │
+│ 2   │ baz    │ 1     │ 6     │
+
+julia> get(gd, (:qux,), nothing)
+```
+"""
+function Base.get(gd::GroupedDataFrame, key::Union{GroupKey, Tuple, NamedTuple}, default)
+    try
+        return gd[key]
```

Comment:
Given this code we should make sure we throw `KeyError` if and only if `key` was not found in `gd` (I assume this is the case but please double check as you know the code best).

Reply:
That should be what it does (not in `groupvars(gd)`), what's the alternative?

Reply:
I think this is what it does, but I simply meant to have a second look at it 😄.

Reply:
There's a test that checks for a KeyError in this case, so ti should be good.

Reply:
But do you test the case where an incorrect `GroupKey` is passed? It seems that that test must fail (`gd[key]` for `GroupKey` does not seem to throw `KeyError`). If you intend for it to work this way then maybe `GroupKey` should not be allowed in this definition?

Reply:
Removed `GroupKey` from the signature